velox/expression/PeeledEncoding.cpp

```diff
@@ -0,0 +1,306 @@
+      // exposed a null complex constant as the base.
+      VELOX_CHECK(peeledVectors.size() == 1);
+      auto innerIdx = decoded->index(rows.begin());
+      VELOX_CHECK(peeledVectors.back()->isNullAt(innerIdx));
+      wrapEncoding_ = VectorEncoding::Simple::CONSTANT;
+      constantWrapIndex_ = innerIdx;
+    } else {
+      setDictionaryWrapping(*decoded, rows, *firstWrapper);
+      // Make sure all the constant vectors have at least the same length as the
+      // base vector after peeling. This will make sure any translated rows
+      // point to valid rows in the constant vector.
+      if (baseSize_ > rows.end() && !constantFields.empty()) {
+        for (int i = 0; i < constantFields.size(); ++i) {
+          if (constantFields[i]) {
+            peeledVectors[i] =
+                BaseVector::wrapInConstant(baseSize_, 0, peeledVectors[i]);
+          }
+        }
+      }
+    }
+  }
+  return true;
+}
+
+VectorEncoding::Simple PeeledEncoding::getWrapEncoding() const {
+  return wrapEncoding_;
+}
+
+VectorPtr PeeledEncoding::wrap(
+    const TypePtr& outputType,
+    velox::memory::MemoryPool* pool,
+    VectorPtr peeledResult,
+    const SelectivityVector& rows) const {
+  VELOX_CHECK(wrapEncoding_ != VectorEncoding::Simple::FLAT);
+  VectorPtr wrappedResult;
+  if (wrapEncoding_ == VectorEncoding::Simple::DICTIONARY) {
+    if (!peeledResult) {
+      // If all rows are null, make a constant null vector of the right type.
+      wrappedResult =
+          BaseVector::createNullConstant(outputType, rows.size(), pool);
+    } else {
+      BufferPtr nulls;
+      if (!rows.isAllSelected()) {
+        // The new base vector may be shorter than the original base vector
+        // (e.g. if positions at the end of the original vector were not
+        // selected for evaluation). In this case some original indices
+        // corresponding to non-selected rows may point past the end of the base
+        // vector. Disable these by setting corresponding positions to null.
+        nulls = AlignedBuffer::allocate<bool>(rows.size(), pool, bits::kNull);
+        // Set the active rows to non-null.
+        rows.clearNulls(nulls);
+        if (wrapNulls_) {
+          // Add the nulls from the wrapping.
+          bits::andBits(
+              nulls->asMutable<uint64_t>(),
+              wrapNulls_->as<uint64_t>(),
+              rows.begin(),
+              rows.end());
+        }
+        // Reset nulls buffer if all positions happen to be non-null.
+        if (bits::isAllSet(
+                nulls->as<uint64_t>(), 0, rows.end(), bits::kNotNull)) {
+          nulls.reset();
+        }
+      } else {
+        nulls = wrapNulls_;
+      }
+      wrappedResult = BaseVector::wrapInDictionary(
+          std::move(nulls), wrap_, rows.end(), std::move(peeledResult));
+    }
+  } else {
+    wrappedResult = BaseVector::wrapInConstant(
+        rows.size(), constantWrapIndex_, std::move(peeledResult));
+  }
```

Comment:
It looks like different paths may produce vectors of different sizes, i.e., rows.size() vs. rows.end(). Would this be a concern? I feel it would be the best to make them consistent.